arangod/Agency/Supervision.cpp

```diff
@@ -3496,3 +3502,207 @@ void Supervision::removeTransactionBuilder(
+    if (not servers) {
+      return false;
+    }
+    TRI_ASSERT(servers->isArray() && servers->length() > 0);
+    for (size_t i = 0; i < servers->length(); ++i) {
+      if (server == servers->at(i).stringView()) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  auto const isServerInSync =
+      [&](std::string_view database, std::string_view collection,
+          std::string_view shard, std::string_view server) -> bool {
+    auto path = basics::StringUtils::joinT("/", "Current/Collections", database,
+                                           collection, shard, "servers");
+    auto servers = snapshot().hasAsArray(path);
+    if (not servers) {
+      return false;
+    }
+    TRI_ASSERT(servers->isArray() && servers->length() > 0);
+    for (size_t i = 1; i < servers->length(); ++i) {
+      if (server == servers->at(i).stringView()) {
+        return true;
+      }
+    }
+    return false;
+  };
+
+  auto const checkDeletion = [&](std::string const& shard,
+                                 std::shared_ptr<Node> entry) -> bool {
+    std::string now(timepointToString(std::chrono::system_clock::now()));
+    auto deadline = entry->hasAsString("removeIfNotStartedBy");
+    auto started = entry->hasAsString("started");
+    if (deadline) {
+      if (!started && now > deadline.value()) {
+        return true;
+      }
+    }
+    if (started) {
+      auto jobId = entry->hasAsString("jobId");
+      if (jobId) {
+        auto inTodo = _snapshot->hasAsNode(toDoPrefix + jobId.value());
+        auto inPending = _snapshot->hasAsNode(pendingPrefix + jobId.value());
+        if (!inTodo && !inPending) {
+          return true;
+        }
+      } else {
+        return true;
+        // This should not happen, since if started is there, we have a
+        // jobId.
+      }
+    }
+    auto jobOpt = entry->hasAsNode("moveShard");
+    if (!jobOpt) {
+      return true;
+    }
+    Node const& job(jobOpt.value());
+    auto database = job.hasAsString("database");
+    auto collection = job.hasAsString("collection");
+    auto server = job.hasAsString("fromServer");
+    if (!server) {
+      return true;
+    } else {
+      if (isReplication2(*database)) {
+        auto stateId = LogicalCollection::shardIdToStateId(shard);
+        auto target = Job::readStateTarget(snapshot(), *database, stateId);
+        if (not target->participants.contains(*server)) {
+          return true;
+        }
+      } else {
+        if (!isServerInPlan(database.value(), collection.value(), shard,
+                            server.value())) {
+          return true;
+        }
+      }
+    }
+    return false;
+  };
+
+  auto const reclaimShard =
+      [&](std::shared_ptr<VPackBuilder> const& trx, std::string const& database,
+          std::string const& collection, std::string const& shard,
+          std::string const& newLeader, std::string const& oldLeader) {
+        uint64_t jobId = _jobId++;
```

Comment:
According to the header file documentation `_jobId` must be guarded by `_lock`, so this could be a potential data race.

Reply:
On a second look, this is actually only called from `handleJobs`, which asserts that it holds the lock. So we have this particular lock for the duration of the Supervision run... So technically, this is all in order. Whether we should hold this lock for so long is a different matter, but this is for a different PR...

Reply:
Ok, as long as it doesn't introduce a potential race, I am fine with it.